static/app/views/organizationContext.tsx

```diff
@@ -1,5 +1,150 @@
+import getRouteStringFromRoutes from 'sentry/utils/getRouteStringFromRoutes';
+import useApi from 'sentry/utils/useApi';
+import {useParams} from 'sentry/utils/useParams';
+import {useRoutes} from 'sentry/utils/useRoutes';
 
 export const OrganizationContext = createContext<Organization | null>(null);
+
+interface Props {
+  children: React.ReactNode;
+}
+
+/**
+ * There are still a number of places where we consume the legacy organization
+ * context. So for now we still need a component that provides this.
+ */
+class LegacyOrganizationContextProvider extends Component<{value: Organization | null}> {
+  static childContextTypes = {
+    organization: SentryPropTypeValidators.isOrganization,
+  };
+
+  getChildContext() {
+    return {organization: this.props.value};
+  }
+
+  render() {
+    return this.props.children;
+  }
+}
+
+/**
+ * Context provider responsible for loading the organization into the
+ * OrganizationStore if it is not already present.
+ */
+export function OrganizationContextProvider({children}: Props) {
+  const api = useApi();
+  const configStore = useLegacyStore(ConfigStore);
+
+  const {organizations} = useLegacyStore(OrganizationsStore);
+  const {organization, error} = useLegacyStore(OrganizationStore);
+
+  const hasMadeFirstFetch = useRef(false);
+
+  const lastOrganizationSlug: string | null =
+    configStore.lastOrganization ?? organizations[0]?.slug ?? null;
+
+  const routes = useRoutes();
+  const params = useParams<{orgId?: string}>();
+
+  // XXX(epurkhiser): When running in deploy preview mode customer domains are
+  // not supported correctly. Do NOT use the customer domain from the params.
+  const orgSlug = DEPLOY_PREVIEW_CONFIG
+    ? lastOrganizationSlug
+    : params.orgId || lastOrganizationSlug;
+
+  // If the organization slug differs from what we have in the organization
+  // store reload the store
+  useEffect(() => {
+    // Nothing to do if we already have the organization loaded
+    if (organization && organization.slug === orgSlug) {
+      return;
+    }
+
+    if (!orgSlug) {
+      return;
+    }
+
+    metric.mark({name: 'organization-details-fetch-start'});
+    fetchOrganizationDetails(api, orgSlug, false, hasMadeFirstFetch.current);
+    hasMadeFirstFetch.current = true;
+  }, [api, orgSlug, organization]);
+
+  // Take a measurement for when organization details are done loading and the
+  // new state is applied
+  useEffect(
+    () => {
+      if (organization === null) {
+        return;
+      }
+
+      metric.measure({
+        name: 'app.component.perf',
+        start: 'organization-details-fetch-start',
+        data: {
+          name: 'org-details',
+          route: getRouteStringFromRoutes(routes),
+          organization_id: parseInt(organization.id, 10),
+        },
+      });
+    },
+    // Ignore the `routes` dependency for the metrics measurement
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [organization]
```

Comment:
Is `organization` stable or should this be organization.id?

Reply:
It should be stable